simulatingrisk/hawkdove/model.py

```diff
@@ -0,0 +1,375 @@
+
+        self.datacollector = mesa.DataCollector(**self.get_data_collector_options())
+
+    def get_data_collector_options(self):
+        # method to return options for data collection,
+        # so subclasses can modify
+        return {
+            "model_reporters": {
+                "max_agent_points": "max_agent_points",
+                "percent_hawk": "percent_hawk",
+                "rolling_percent_hawk": "rolling_percent_hawk",
+                "status": "status",
+                # explicitly track total agents, instead of inferring from grid size
+                "total_agents": "num_agents",
+            },
+            "agent_reporters": {
+                "risk_level": "risk_level",
+                "choice": "choice_label",
+                "points": "points",
+            },
+        }
+
+    def new_agent_options(self):
+        # generate and return a dictionary with common options
+        # for initializing all agents
+        return {"hawk_odds": self.hawk_odds}
+
+    def step(self):
+        """
+        A model step. Used for collecting data and advancing the schedule
+        """
+        self.schedule.step()
+        # check if simulation has converged and should stop running
+        if self.converged:
+            self.status = "converged"
+            self.running = False
+
+        # collect data after status is updated, so data collected
+        # for last round will reflect converged status
+        self.datacollector.collect(self)
+
+    @property
+    def max_agent_points(self):
+        # what is the current largest point total of any agent?
+        return max([a.points for a in self.schedule.agents])
+
+    @property
+    def percent_hawk(self):
+        # what percent of agents chose hawk?
+        hawks = [a for a in self.schedule.agents if a.choice == Play.HAWK]
+        phawk = len(hawks) / self.num_agents
+        # add to recent values
+        self.recent_percent_hawk.append(phawk)
+        return phawk
+
+    @property
+    def rolling_percent_hawk(self):
+        # make sure we have enough values to check
+        if len(self.recent_percent_hawk) > self.min_window:
+            rolling_phawk = statistics.mean(self.recent_percent_hawk)
+            # add to recent values
+            self.recent_rolling_percent_hawk.append(rolling_phawk)
+            return rolling_phawk
+
+    @property
+    def converged(self):
+        # check if the simulation is stable and should stop running
+        # calculating based on rolling percent hawk; when this is stable
+        # within our rolling window, return true
+        # - currently checking for single value;
+        #  could allow for a small amount variation if necessary
+
+        # in variable risk with risk adjustment, numbers are not strictly equal
+        # but do get close and fairly stable; round to two digits before comparing
+        rounded_set = set([round(x, 2) for x in self.recent_rolling_percent_hawk])
```

Comment:
For my understanding: In the deck of recent phawk values, round all values to two digits, and compare if they are still different. If we have 15 values to compare and all of them are the same after rounding, this is considered converged.